### PR TITLE
Extract preprocessing from detector

### PR DESCRIPTION
**Pull request by janwestphal306**
Extract preprocessing from detector and test it.

### CODE REVIEW

File: shift_detector/Detector.py [hunk 1]
```diff
@@ -28,14 +28,14 @@ def __init__(self,
         if type(df1) is pd.DataFrame:
```
Comment (phSch08) [inline shift_detector/Detector.py]: What about this TODO

File: shift_detector/Detector.py [hunk 2]
```diff
@@ -52,53 +52,6 @@ def add_checks(self, checks: List[Check]):
...
-
-        for column_type, needed_preprocessings in type_to_needed_preprocessings.items():
-            (first_df, second_df) = column_type_to_columns[column_type]
-            for needed_preprocessing in needed_preprocessings:
-                preprocessed = needed_preprocessing.process(first_df, second_df)
-                preprocessings[column_type][needed_preprocessing] = preprocessed
-
-        return preprocessings
-
     def _distribute_preprocessings(self, checks: List[Check], preprocessings: Dict):
```
Comment (phSch08) [inline shift_detector/Detector.py]:
```suggestion
    @staticmethod
    def _distribute_preprocessings(self, checks: List[Check], preprocessings: Dict):
```

this method can be declared as static

File: shift_detector/Utils.py [hunk 3]
```diff
@@ -12,83 +12,80 @@ class ColumnType(Enum):
...
-        text_columns = list(set(columns) - set(numeric_columns) - set(categorical_columns))
-        logger.info("Assuming text columns: {}".format(", ".join(text_columns)))
-
-        return {
-            ColumnType.numeric: (df1[numeric_columns], df2[numeric_columns]),
-            ColumnType.categorical: (df1[categorical_columns], df2[categorical_columns]),
-            ColumnType.text: (df1[text_columns], df2[text_columns])
-        }
+def read_from_csv(file_path: str, separator) -> pd.DataFrame:
+    # TODO: give user feedback about how many lines were dropped
```
Comment (phSch08) [inline shift_detector/Utils.py]: Here is still a TODO

File: shift_detector/Utils.py [hunk 4]
```diff
@@ -12,83 +12,80 @@ class ColumnType(Enum):
...
-        logger.info("Assuming categorical columns: {}".format(", ".join(categorical_columns)))
-        text_columns = list(set(columns) - set(numeric_columns) - set(categorical_columns))
-        logger.info("Assuming text columns: {}".format(", ".join(text_columns)))
-
-        return {
-            ColumnType.numeric: (df1[numeric_columns], df2[numeric_columns]),
-            ColumnType.categorical: (df1[categorical_columns], df2[categorical_columns]),
-            ColumnType.text: (df1[text_columns], df2[text_columns])
-        }
+def read_from_csv(file_path: str, separator) -> pd.DataFrame:
```
Comment (phSch08) [inline shift_detector/Utils.py]:
```suggestion
def read_from_csv(file_path: str, separator: str) -> pd.DataFrame:
```
for consistency reasons

File: shift_detector/Utils.py [hunk 5]
```diff
@@ -12,83 +12,80 @@ class ColumnType(Enum):
...
+    # TODO: give user feedback about how many lines were dropped
+    logger.info('Reading in CSV file. This may take a while ...')
+    return pd.read_csv(file_path, sep=separator, error_bad_lines=False).dropna()
+
+
+def shared_column_names(df1: pd.DataFrame, df2: pd.DataFrame) -> List[str]:
+    """
+    Find the column names that both dataframes share.
+    Raise an exception if the dataframes do not a shared column name.
```
Comment (phSch08) [inline shift_detector/Utils.py]:
```suggestion
    Raise an exception if the dataframes do not have a shared column name.
```

File: shift_detector/Utils.py [hunk 6]
```diff
@@ -12,83 +12,80 @@ class ColumnType(Enum):
...
+                            "They have {} and {}".format(df1_columns, df2_columns))
+    else:
+        shared_columns = df1_columns
+
+    return list(shared_columns)
+
+
+def is_categorical(col: pd.Series,
+                   n_samples: int = 100,
+                   max_unique_fraction=0.1) -> bool:
```
Comment (phSch08) [inline shift_detector/Utils.py]:
```suggestion
                   max_unique_fraction: float = 0.1) -> bool:
```

File: shift_detector/Utils.py [hunk 7]
```diff
@@ -12,83 +12,80 @@ class ColumnType(Enum):
...
+        if len(shared_columns) == 0:
+            raise Exception("The provided datasets do not have any column names in common."
+                            "They have {} and {}".format(df1_columns, df2_columns))
+    else:
+        shared_columns = df1_columns
+
+    return list(shared_columns)
+
+
+def is_categorical(col: pd.Series,
```
Comment (phSch08) [inline shift_detector/Utils.py]: is it reasonable to test this with a relative cardinality?

File: shift_detector/preprocessors/Preprocessing.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+import pandas as pd
+
+from shift_detector.Utils import split_dataframes
+from shift_detector.checks.Check import Check
+
+
+def preprocess(checks: List[Check],
+               df1: pd.DataFrame,
+               df2: pd.DataFrame,
+               columns: List[str]):
```
Comment (phSch08) [inline shift_detector/preprocessors/Preprocessing.py]:
```suggestion
               columns: List[str]) -> Dict:
```

File: shift_detector/Detector.py (hunk 1 above)
Comment (janwestphal306) [reply shift_detector/Detector.py]: Speeds up debugging but needs to be removed later

File: shift_detector/Utils.py (hunk 3 above)
Comment (janwestphal306) [reply shift_detector/Utils.py]: Maybe will be added in the future

**Comment by janwestphal306** [comment]
Detector and Utils will receive more testing in later PRs.

**Pull request merged**